feat(providers): Add server-side encryption support for S3 storage

apampurin opened the pull request:
# Add Server-Side Encryption (SSE) Support for S3 Storage

## Description
This PR adds support for Server-Side Encryption (SSE) in S3 storage, allowing users to encrypt their data at rest using either AES-256 or AWS KMS encryption methods. This enhances data security by ensuring that data is encrypted before being stored in S3 buckets.

## Features
- Support for AES-256 server-side encryption
- Support for AWS KMS server-side encryption with:
  - Default AWS KMS key
  - Custom KMS key (when key ID is provided)
- Comprehensive test coverage for all encryption methods

## Implementation Details
- Added `ServerSideEncryption` and `KMSKeyID` options to S3 storage configuration
- Implemented encryption configuration in the S3 client
- Added test cases for different encryption scenarios:
  - AES-256 encryption
  - AWS KMS with default key
  - AWS KMS with custom key (optional)

## Testing
Added test cases in `TestS3StorageEncryption` that verify:
- AES-256 encryption works correctly
- AWS KMS encryption works with default key
- AWS KMS encryption works with custom key (when provided)
- Proper error handling for missing KMS key ID

## Usage
To use server-side encryption, configure the S3 storage with the desired encryption method:

```go
options := &Options{
    // ... other options ...
    ServerSideEncryption: "AES256",  // or "aws:kms" for KMS encryption
    KMSKeyID: "arn:aws:kms:...",    // optional, only needed for custom KMS key
}
```

## Security Considerations
- Encryption is handled server-side by AWS S3
- For KMS encryption, proper IAM permissions are required
- Custom KMS keys require appropriate key policies and IAM permissions

## Dependencies
- No new dependencies added
- Uses existing AWS SDK capabilities

## Related Issues
Closes #4570 - Add server-side encryption support for S3 storage

Commit:
feat(s3): add server-side encryption options for S3 storage

- Introduced `ServerSideEncryption` and `KMSKeyID` fields in the Options struct to support server-side encryption.
- Implemented tests for AES256 and AWS KMS encryption options in S3 storage.
- Updated the `putBlob` method to configure server-side encryption during object uploads.

Commit:
chore(go.mod): update module path from github.com/kopia/kopia to github.com/apampurin/kopia

Commit:
feat(s3): enhance server-side encryption support in S3 storage

- Updated the `putBlob` method to support multiple server-side encryption methods: AES256 and AWS KMS.
- Integrated new encryption options from the MinIO Go SDK for better flexibility in handling encryption during uploads.
- Added error handling for unsupported encryption methods.

Commit:
fix(s3): add error handling for KMS encryption initialization

- Enhanced the `putBlob` method to include error handling when initializing KMS encryption.
- Ensured that any errors during the setup of server-side encryption are properly wrapped and returned.

Commit:
feat(s3): add server-side encryption flags for S3 storage

- Introduced new command-line flags for server-side encryption: `--server-side-encryption` and `--kms-key-id`.
- These flags allow users to specify the encryption method and KMS key ID when configuring S3 storage options.

Commit:
refactor(s3): update S3 storage encryption handling and fix module path

- Refactored the `putBlob` method to streamline server-side encryption configuration, improving readability and maintainability.
- Enhanced the handling of server-side encryption options during object uploads, ensuring proper initialization and error handling.

Commit:
refactor(s3): simplify putBlob method by removing redundant comments

- Cleaned up the `putBlob` method in the S3 storage implementation by removing unnecessary comments related to content type and multipart uploads.
- Streamlined the code for handling empty streams while maintaining the existing functionality and configuration options.

Commit:
return to base view

julio-lopez commented on repo/blob/s3/s3_storage.go:
```diff
@@ -201,6 +225,29 @@ func (s *s3Storage) putBlob(ctx context.Context, b blob.ID, data blob.Bytes, opt
+				case "AES256":
+					return encrypt.NewSSE()
+				case "aws:kms":
+					var err error
+					var sse encrypt.ServerSide
+					if s.KMSKeyID != "" {
+						sse, err = encrypt.NewSSEKMS(s.KMSKeyID, nil)
+					} else {
+						sse, err = encrypt.NewSSEKMS("", nil)
+					}
```
same as above

julio-lopez commented on repo/blob/s3/s3_storage.go:
```diff
@@ -182,6 +183,29 @@ func (s *s3Storage) putBlob(ctx context.Context, b blob.ID, data blob.Bytes, opt
+				} else {
+					sse, err = encrypt.NewSSEKMS("", nil)
+				}
+				if err != nil {
```
This is silently ignoring the error, is that intended?

apampurin replied:
es, you're right about the error handling. I've moved the encryption logic to a new getServerSideEncryption() function which properly handles and propagates errors

julio-lopez reviewed (commented):
This is going in the right direction.

- error handling is needed when putting blobs, in particular it should fail early if there are any issues setting up SSE.
- perhaps the SSE setup could be done only once when a `s3Storage` instance is created.
- the implementation, the interface and the naming should be such that it also works with other S3-compatible stores that implement SSE.

Commit:
feat(s3): enhance server-side encryption configuration and validation

- Added a new command-line flag for server-side encryption method with validation for accepted values (NONE, AES256, KMS).
- Refactored the `putBlob` method to utilize a dedicated function for retrieving server-side encryption settings, improving code clarity and maintainability.
- Updated tests to ensure proper handling of server-side encryption options during S3 storage operations.

apampurin commented on repo/blob/s3/s3_storage.go:
```diff
@@ -182,6 +183,29 @@ func (s *s3Storage) putBlob(ctx context.Context, b blob.ID, data blob.Bytes, opt
+					sse, err = encrypt.NewSSEKMS(s.KMSKeyID, nil)
+				} else {
+					sse, err = encrypt.NewSSEKMS("", nil)
+				}
```
Thank you for the suggestion. You're right - the simplified version is cleaner and achieves the same result.

apampurin commented on repo/blob/s3/s3_storage.go:
```diff
@@ -201,6 +225,29 @@ func (s *s3Storage) putBlob(ctx context.Context, b blob.ID, data blob.Bytes, opt
+				default:
+					return nil
+				}
+			}(),
```
Done.

Commit:
feat(s3): enhance S3 storage tests for server-side encryption

- Added comprehensive tests for server-side encryption in S3 storage, including support for AES256 and AWS KMS.
- Implemented a new function to start a MinIO instance with static key encryption for testing purposes.
- Created bucket policies that enforce server-side encryption for object uploads, validating both successful and denied upload scenarios based on encryption headers.
- Improved documentation within the test file to clarify the purpose and functionality of the tests.

SCLogo commented:
any updates on this?
